subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,565 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/instrumentation/instrumentation.h>
+#include <zephyr/instrumentation/instrumentation_timestamp.h>
+#include <zephyr/instrumentation/instrumentation_ringbuffer.h>
+#include <zephyr/kernel.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
+
+#include <zephyr/sys/reboot.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+
+#include <zephyr/drivers/retained_mem.h>
+
+#include <kernel_internal.h>
+#include <ksched.h>
+
+/* Memory buffer to store instrumentation event records has the following modes:
+
+ - Callgraph ring buffer (default): Replace oldest entry when buffer
+ full, at the cost of processing stray entry events that have been
+ truncated.
+
+ - Callgraph fixed buffer: Stop buffering events when the buffer is
+ full, ensuring we have a callgraph from reset or from whenever the
+ trigger event was entered.
+
+ - Statistical: Buffer functions until out of memory
+
+*/
+
+const static struct device *retained_mem_dev = DEVICE_DT_GET(DT_ALIAS(retainedmemdevice));
```

Comment:
Would use a chosen node for this, similar to this: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/boot/mcuboot_recovery_retention/boards/nrf52840dk_nrf52840_mem.overlay

Reply:
Done.

Reply:
@nordicjm `chosen` in this example is select boot mode, which is not really necessary, but I'm not sure I'm following you here. Could you please expand on it?

Reply:
In a board dts or an overlay, you could have something like this:
```
	retainedmem {
		compatible = "zephyr,retained-ram";
		status = "okay";
		#address-cells = <1>;
		#size-cells = <1>;

		instru0: instru@0 {
			compatible = "zephyr,retention";
			status = "okay";
			reg = <0x0 0x100>;
			prefix = [08 04];
			checksum = <1>;
		};
	};

	chosen {
		zephyr,instrumentation-area = &instru0;
	};
```
Then that way you can access it by using `static const struct device *retained_mem_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_instrumentation_area));`
This is not a required change.

Reply:
ah, got it. Thanks for the explanation!